### ISSUE TITLE
Support for Ubuntu 20.04 focal

### ISSUE THREAD

**Issue by mprasil** [zerotier/install.zerotier.com#13]
This Ubuntu version is going to be released in couple months, daily builds are already available. Maybe we could get support in before it's officially out?

**Comment by jdevora** [comment]
I upgraded to 20.04 and when I try apt-get update it gives me the error:
```
Err http://download.zerotier.com/debian/bionic eoan Release
  404  Not Found [IP: 35.244.178.109 80]
E: The repository 'http://ppa.launchpad.net/bitcoin/bitcoin/ubuntu focal Release' does not have a Release file.
E: The repository 'http://download.zerotier.com/debian/bionic eoan Release' does not have a Release file.
W: Failed to fetch http://download.zerotier.com/debian/bionic/dists/eoan/Release: 404  Not Found [IP: 35.244.178.109 80]

```

**Comment by GMDSantana** [comment]
> I upgraded to 20.04 and when I try apt-get update it gives me the error:
> 
> ```
> Err http://download.zerotier.com/debian/bionic eoan Release
>   404  Not Found [IP: 35.244.178.109 80]
> E: The repository 'http://ppa.launchpad.net/bitcoin/bitcoin/ubuntu focal Release' does not have a Release file.
> E: The repository 'http://download.zerotier.com/debian/bionic eoan Release' does not have a Release file.
> W: Failed to fetch http://download.zerotier.com/debian/bionic/dists/eoan/Release: 404  Not Found [IP: 35.244.178.109 80]
> ```

@jdevora Try using "http://download.zerotier.com/debian/bionic bionic" rather than "http://download.zerotier.com/debian/bionic eoan"

**Comment by Timple** [comment]
But 20.04 is not **bionic** right? It should be **focal**

**Comment by mprasil** [comment]
Focal has been out for couple months and it's the current latest LTS Ubuntu release. Can we get some update for this?

**Comment by ghaschel** [comment]
Zerotier is misbehaving a lot lately in 20.04. It sometimes work, sometimes doesn't.

It bugs a lot in a machine with a clean 20.04 install. Doesn't bug at all in a machine which was upgraded from 19 to 20.

**Comment by Halfwalker** [comment]
Still waiting for a 20.04/focal repo and release.  My ansible playbooks have needless complexity to accommodate a focal system with a non-focal (bionic) zerotier release.

**Comment by asper-ru** [comment]
Still nothing new?

I fix installation parameters to use ubuntu 19.04 repo, but it's strange to do not support Ubuntu 20 yet, almost in 2021

**Comment by ghaschel** [comment]
I've been using it on Ubuntu 20.04 without problems so far.

**Comment by TeoTN** [comment]
It seems it might be better to use Nebula instead :/

**Comment by nukebloodaxe** [comment]
@asper-ru I suspect most of their resources are allocated to the version 2 release, and this current release is not being maintained as closely as it should be.  Given new customers before the 2.0 release will be attempting to use this version, I believe they should at least fix these install issues so people can see it is a product they want to buy; something which does not install will not inspire confidence in a business researcher, and will likely stop them from looking again, even if 2.0 comes out, as hey will have moved on to testing competing products [simple psychology.]

**Comment by ukd1** [comment]
This is kind of a joke it's broken for this long. Are y'all supporting this at all anymore?

**Comment by einverne** [comment]
Still waiting for the support for 20.04

**Comment by sebzur** [comment]
@mprasil Any chances for 20.04?

**Comment by sebzur** [comment]
OK, I've found snap option: https://snapcraft.io/zerotier - it worked. However, under https://www.zerotier.com/download/ there's no mention about it. Is it okay or rather one should use `curl|bash` installation (it seems to be the preferred way: https://github.com/zerotier/install.zerotier.com)

**Comment by mprasil** [comment]
To quote myself from 2y ago: 

> This Ubuntu version is going to be released in couple months, daily builds are already available. Maybe we could get support in before it's officially out?

There are already [22.04 daily builds available](https://cdimage.ubuntu.com/daily-live/current/). The fact that there's still no 20.04 (focal) support in the install script is a bit sad, is there any PR I could submit somewhere to help you resolve that?

**Comment by alyandon** [comment]
@mprasil I too have been waiting for a while now and it really seems to me like this is a dead sub-project.  The last commit in this repository was in 2019 and PRs that were opened to clean up and improve the installer have been open since 2018.  If ZT isn't actually building and packaging .debs for newer Ubuntu releases then it seems like building from source is the only viable solution.

**Comment by adamierymenko** [comment]
It's not abandoned, but it hasn't been updated often. We can test on focal.

**Comment by adamierymenko** [comment]
A quick test on focal seems to work. Can someone give more information?

**Comment by joseph-henry** [comment]
> OK, I've found snap option: https://snapcraft.io/zerotier - it worked. However, under https://www.zerotier.com/download/ there's no mention about it. Is it okay or rather one should use `curl|bash` installation (it seems to be the preferred way: https://github.com/zerotier/install.zerotier.com)?

Yes. This works on Focal. Though there's a known issue with Snap's network security policy that prevents outbound connections sometimes. I'll have to speak with them about how to resolve that.

And to expand further on what adam said, this isn't abandoned. Part of our team uses this on latest Ubuntu as daily drivers. Now we could do a better job making official releases for each Ubuntu version, that much is true.

**Comment by adamierymenko** [comment]
In the past we have often defaulted to using the most recent named release on newer releases if it works fine, so that explains "focal" getting "bionic" etc. You do not necessarily need to build new binaries for every single release.

**Comment by alyandon** [comment]
Interesting, I just provisioned a Ubuntu 20.04.3 LTS VM and ran the script.  It appeared to complete successfully and generated the node ID.

However, the end result is the following in /etc/apt/sources.list.d/zerotier.list

```
deb http://download.zerotier.com/debian/bionic bionic main
```

So it looks like it installed the bionic package on focal - which probably works only because 18.04 vs 20.04 are similar enough.  Additionally, http://download.zerotier.com/debian/focal appears to not exist which implies ZT isn't building and publishing a focal package.

**Comment by adamierymenko** [comment]
Yeah we're not yet. We haven't a need to add it unless the older package has issues on focal. Maybe we should use some kind of generic naming convention to keep people from seeing a different distribution name in there and thinking it means something.

**Comment by alyandon** [comment]
If it works and is supported then I'm find.  For what it is worth - I don't envy you having to package things for a bazillion different distros - it's a lot of work.

**Comment by adamierymenko** [comment]
We package things for N distributions times many different architectures. It is a lot of work and makes builds very slow. There's no need to add new ones when it's not necessary, but I will give some thought to generic "revision categories" or something similar. Not sure exactly how to solve this issue. It makes supporting Linux really painful sometimes.

**Comment by mprasil** [comment]
You could also look into discontinuing support for EOL Ubuntu releases like `zesty`, `artful` or `precise`. These have been EOL for years now. If can help and free up some bandwidth to properly support current LTS versions, I'd say it would be a worthy compromise.

**Comment by alyandon** [comment]
@mprasil Based on the discussion, it seems like they are just names in the script and don't really reflect support for a particular Ubuntu release.  Just like `bionic` in the repo URL is apparently not specifically meant to imply support for only `bionic` - it was merely the latest LTS when they wrote the script and ZT moving to a generic repo name would fix that confusion.

Did you see above where I ran the installer script on a fresh 20.04 VM?  Are you still getting the same error you reported?

**Comment by mprasil** [comment]
So I'm actually not using these scripts at all. There are places where I deploy in more automated fashion like with Ansible and it's just pain to have exception for specific releases, because there is no package in the repository for specific distribution version.

For example (this is a role that I'm not actually using, but it demonstrates the problem well) this [code here](https://github.com/m4rcu5nl/ansible-role-zerotier/blob/ca3cc077d695fe7d7dc703bb74b910d1bef4adc8/tasks/install/Debian.yml#L6-L25) is just to work around the problem.

Now granted, this is not a huge problem considering the package otherwise works well on the system, but it looks quite confusing. (even more so in the apt configuration file that ends up on the machine) And considering Ubuntu 20.04 is probably the number 1 most used distribution release in the cloud (or at least near the top) it looks weird that this kind of release isn't directly supported. Especially if it sounds like some alias in Debian repository is all that's really needed.

When I originally reported the issue, 20.04 wasn't yet released, so I reported this with hope that this will be resolved before that LTS version was out and added a workaround "for the time being". I could imagine some new users might try to just add the repo, get an 404 for their distribution and move on to some other product thinking it's unsupported.

**Comment by sebzur** [comment]
> When I originally reported the issue, 20.04 wasn't yet released, so I reported this with hope that this will be resolved before that LTS version was out and added a workaround "for the time being". I could imagine some new users might try to just add the repo, get an 404 for their distribution and move on to some other product thinking it's unsupported.

And that precisely was my first thought... The project is not well supported... What a pity.. The status of this official recommended installation procedure for U20.04 have not only surprised me but disappointed as well. I'm at the moment of important decision - ZeroTier or WireGuard? The two are the sexiest candidates to upgrade the security of our huge network of distributed end-points. I love the idea behind ZT. That's why I decided to give it a second chance and experiment with `snap` (which after so many years with `apt` is still somehow unfamiliar for me...) Anyway - do you remember that big title on your front-page?

![obraz](https://user-images.githubusercontent.com/1037150/153504944-884f2230-5c86-4423-b4e5-cd641c4c1de1.png)
    
(Sorry for beeing so blunt - but I - and I guess everybody - would love if the ZT in U20.04 will **just work**)

But... But wait.. @alyandon - have you changed anything lately... I just wanted to experiment with your answer here: https://github.com/zerotier/install.zerotier.com/issues/13#issuecomment-1026267490 - but before that, on my another U20.04 device I first tried `curl -s https://install.zerotier.com | sudo bash`... And... **It just worked!**. Hurrra! 

![obraz](https://user-images.githubusercontent.com/1037150/153502946-d18a39cf-c2b4-44eb-a297-c067efc33eae.png)

@alyandon any comments on that?

**Comment by alyandon** [comment]
@sebzur I didn't change anything.  I provisioned an Ubuntu 20.04.3 VM and ran the very same script you ran which resulted in a working install of ZT for me.

**Comment by sebzur** [comment]
@alyandon - so probably this is somehow related to this comment: https://github.com/zerotier/install.zerotier.com/issues/13#issuecomment-668595790

On my `home-office` host I have Ubuntu upgraded multiple times for a few years. Not sure what wash the first U version there... My first comment here https://github.com/zerotier/install.zerotier.com/issues/13#issuecomment-1018994233 was related to the fault installation on my laptop - I bought it something like a year or so ago. Chances are that I had 20 installed there for a first time.

**Comment by sebzur** [comment]
@alyandon - thank you for your engagement. During the analysis of `curl -s https://install.zerotier.com | sudo bash -x` I've noticed that I had local problem with some missing pubkeys (NO_PUBKEY GPG error). I should have noticed it earlier! I was trying to add the keys manually (i.e. for ZT using `sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 1657198823E52A61`) which failed... 

After some investigation I found, that there was some problematic gpg key (the error said `gpg: invalid key resource URL '/tmp/apt-key-gpghome.GOWinQfCit/home:cabelo.asc.gpg'`) - I've simply deleted it (`sudo rm /etc/apt/trusted.gpg.d/home\:cabelo.asc`), then again processed `sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 1657198823E52A61` which succeeded.

After that, again standard ZT installation step with `sudo curl -s https://install.zerotier.com | sudo bash` ... and it worked as promised on ZT front page :exclamation: :heart: 

So it looks (at least in my case) that the only problematic case was my inattention and some mess in the system. Sorry for this dramatic post here https://github.com/zerotier/install.zerotier.com/issues/13#issuecomment-1035588764 And thank you again for the whole work here with the ZT

Anyway - to summarize my problem and leave something potentially useful for someone -  if you got stuck at that installation step, check:

- 1) are there any NO_PUBKEY errors during the instllation of ZT?
- 2) if so, you can execute `sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys ZT_KEY_IDENTIFIER` (be aware what key you are adding)
- 3) in case of errors like `gpg: invalid key resource URL '/tmp/apt-key-gpghome.GOWinQfCit/home:cabelo.asc.gpg'` remove the problematic key `rm /etc/apt/trusted.gpg.d/{Problem-Key}.asc`
- execute step 2) again, and then you should be able to install ZT with suggested `curl | bash` method

**Comment by alyandon** [comment]
Oh cool - glad you were able to get that working.  ZT probably needs to update the installer script to validate the repo key is actually accessible via GPG.

**Comment by mprasil** [comment]
Not sure when the change happened (probably sometimes end of April looking at the file dates), but now I can see there is repository for 20.04 (focal) and even for recently released 22.04 (jammy) available. Thanks a lot for that folks, I think we can consider this issue resolved.

Whoever worked on this, you made my life a little bit easier, thank you again.
